### PR TITLE
feat: add excludes to @CopyMapping / @CombineMapping

### LINKED ISSUE #164: Add excludes to @CopyMapping / @CombineMapping (annotation-level @Exclude for external classes)
## Summary

Add an `excludes` parameter to `@CopyMapping` and `@CombineMapping` — the annotation-level equivalent of `@Exclude` for the library-to-library mapping annotations. Since the source/target classes live in external libraries and cannot be annotated with `@CopyTo.Exclude` / `@CopyFrom.Exclude` etc., `excludes` lets you remove the auto-copy default (making the parameter required) by naming the property directly on the mapping annotation.

## Motivation

`@Exclude` removes a generated parameter's `= this.<property>` default, forcing the caller to supply an explicit value:

| Annotation | Where `@Exclude` is placed |
|---|---|
| `@CopyTo.Exclude` | source class constructor parameter |
| `@CopyFrom.Exclude` | target class constructor parameter |
| `@CombineTo.Exclude` | source class property |
| `@CombineFrom.Exclude` | target class constructor parameter |

`@CopyMapping` / `@CombineMapping` cannot use any of these, because they map **classes you don't own** — there is no property to annotate. This is a known asymmetry (README: "`@CopyMapping` and `@CombineMapping` do not support `@Exclude` … annotating their properties is not possible"; feature × option matrix marks both as ✕ for `.Exclude`).

The parallel gap for **name mapping** was already solved the same way: because you can't put `@CopyTo.Map` on an external property, the mapping annotations accept `properties = [Map(source, target)]` as an annotation argument. `excludes` is simply the `@Exclude` counterpart of that pattern.

## Current state

```kt
public annotation class CopyMapping(
    val source: KClass<*>,
    val target: KClass<*>,
    val canReverse: Boolean = false,
    val properties: Array<Map> = [],   // <- name mapping already exists
    val kdoc: KDoc = KDoc(),
    val funName: String = DefaultCopyFunctionName,
    val visibility: CopyVisibility = CopyVisibility.INHERIT,
) { /* Map(source, target) */ }

public annotation class CombineMapping(
    val sources: Array<KClass<*>>,
    val target: KClass<*>,
    val properties: Array<Map> = [],   // <- name mapping already exists
    /* kdoc / visibility / funName */
) { /* Map(source, target) */ }
```

There is no way to make a matched (auto-defaulted) parameter required.

## Proposal

Add `val excludes: Array<String> = []` to both annotations. Each entry names a **generated parameter** (i.e. a target-side property name) whose auto-copy default should be dropped, mirroring `@Exclude`.

```kt
// in library X / Y (not yours)
data class LibXModel(val shareProp: String, val xProp: Int)
data class LibYModel(val shareProp: String, val yProp: Int)

// in your module
@CopyMapping(
    source = LibXModel::class,
    target = LibYModel::class,
    excludes = ["shareProp"],   // drop the auto-default for shareProp
)
private object Mapping

// auto generate
fun LibXModel.copyToLibYModel(
    shareProp: String,          // no `= this.shareProp` default — now required
    yProp: Int,
): LibYModel = ...
```

### Semantics

- An `excludes` entry that matches a generated (auto-defaulted) parameter removes its `= this.<property>` default; the parameter stays in the signature but becomes required. Identical effect to `@Exclude`.
- An entry that does **not** match any auto-defaulted parameter has no effect and emits a **KSP warning** — same behaviour `@Exclude` already has for unmatched parameters.
- Interaction with `properties = [Map(...)]`: the generated parameter is named after the target, so `excludes` entries reference **target-side names** (consistent with the generated signature). A renamed-then-excluded property is fine (`Map(source="xProp", target="yProp")` + `excludes=["yProp"]`).

### `canReverse` (CopyMapping only)

With `canReverse = true`, the reverse function's parameters are named after the source. `excludes` should apply in both directions, translated through the same `Map` reversal cream already does — for same-named shared properties (the common `@Exclude` case) no translation is needed. The exact reverse-side naming rule is an open question below.

## Scope of changes

- **runtime** `cream-runtime/.../CopyMapping.kt` and `CombineMapping.kt`: add `val excludes: Array<String> = []` + KDoc (`@param excludes`).
- **ksp** `feature/copyMapping/` and `feature/combineMapping/`: read `excludes` from the annotation and thread it into default generation.
- **ksp** `core/common/` (property matching / default generation, e.g. `findMatchedProperty`) + `core/copyFun` / `core/combineFun`: when a matched parameter's name is in `excludes`, emit it without the `= this.<property>` default (required). Reuse the existing `@Exclude` "drop default" path so behaviour is identical.
- **diagnostics**: warn on `excludes` entries that match no auto-defaulted parameter (mirror the existing `@Exclude`-unmatched warning).
- **docs**: README "Exclude" section (add the mapping-annotation form) + update the feature × option matrix (CopyMapping / CombineMapping `.Exclude` row: ✕ → ✓ via `excludes`).
- **tests**: snapshot cases for `@CopyMapping(excludes=[...])`, `@CombineMapping(excludes=[...])`, `canReverse=true` + `excludes`, and the unmatched-name warning; plus `test/` E2E.

## Open questions

- API shape: plain `excludes: Array<String>` (names) vs. a nested `@CopyMapping.Exclude(target = "...")` used as `excludes: Array<Exclude>` (parallels `Map` but heavier). `Array<String>` is simpler since an exclude only needs one name; the nested form buys symmetry with `Map` and room to grow.
- Which side the names reference (target-side proposed here) and the exact `canReverse` translation rule for renamed properties.

## Todo

- [ ] runtime: add `excludes: Array<String>` to `@CopyMapping` and `@CombineMapping` (+ KDoc)
- [ ] ksp: read `excludes` in the copyMapping / combineMapping features
- [ ] ksp: drop the auto-default for excluded parameters (reuse the `@Exclude` path in core)
- [ ] diagnostics: warn on unmatched `excludes` entries
- [ ] docs: README "Exclude" section + feature × option matrix
- [ ] snapshot + E2E tests (incl. `canReverse` + `excludes`, unmatched-name warning)

## References

- README "CopyTo.Exclude … / CopyToChildren.Exclude" section (the `@Exclude` semantics this mirrors).
- Feature × option matrix: `.Exclude` row marks CopyMapping / CombineMapping as unsupported ("対象クラスを編集できないため").
- Precedent: `properties = [Map(source, target)]` — the annotation-level equivalent of `.Map` for the same external-class reason.

### PR DESCRIPTION
**Pull request by TBSten**
## Summary

Adds an `excludes: Array<String> = []` argument to `@CopyMapping` and `@CombineMapping` (issue #164). `@Exclude` removes a generated parameter's `= this.<property>` auto-copy default, but the mapping annotations target classes you don't own, so there is no property to annotate. `excludes` is the annotation-argument counterpart: each entry names a generated parameter whose default is dropped, making it required.

```kt
@CopyMapping(
    source = LibXModel::class,
    target = LibYModel::class,
    excludes = ["shareProp"],
)
private object Mapping

// generates:
fun LibXModel.copyToLibYModel(
    shareProp: String, // required — no default
    yProp: Int,
): LibYModel = ...
```

## Semantics

- **Target-side naming rule**: each `excludes` entry names a generated parameter, i.e. a **target-side** property name — consistent with the generated signature, including `properties = [Map(source, target)]` renames. `Map(source = "xProp", target = "yProp")` + `excludes = ["yProp"]` makes the renamed parameter required.
- **Same effect as `@Exclude`**: the entry is resolved through the existing `isExcludedFromCopy` drop-default path in `core/common/ExcludeProperty.kt`, so the parameter stays in the signature (and is listed as required in the generated KDoc examples) but loses its default.
- **`canReverse = true` translation rule** (CopyMapping only): `excludes` applies in **both** directions. For the reverse function each entry is translated through the reversed `Map` pairs: an entry naming the `target` of a `Map(source, target)` excludes the reverse function's **source-side** parameter; entries without a mapping (same-named shared properties) apply as-is.

  ```kt
  @CopyMapping(
      source = LibXModel::class, target = LibYModel::class,
      canReverse = true,
      properties = [CopyMapping.Map(source = "xProp", target = "yProp")],
      excludes = ["yProp"],
  )
  private object Mapping
  // forward:  fun LibXModel.copyToLibYModel(shareProp = this.shareProp, yProp: Int /* required */)
  // reverse:  fun LibYModel.copyToLibXModel(shareProp = this.shareProp, xProp: Int /* required */)
  ```
- **Unmatched entries**: an entry that suppresses no auto-copy default (typo, or a parameter that had no default anyway) has no effect and emits a KSP **warning** anchored at the mapping holder, mirroring the unmatched `@Exclude` warning. The check is evaluated once per mapping annotation across all directions (see 729f540c): an entry is effective if it suppresses a default in any generated overload (any constructor, any direction, every concrete leaf of a sealed target). An object target generates a parameterless function, so its `excludes` entries can never be effective and always warn — surfacing the pointless entry, per the issue spec ("an entry that does not match any auto-defaulted parameter has no effect and emits a KSP warning").

## Changes

- `cream-runtime`: `excludes` parameter on `@CopyMapping` / `@CombineMapping` + KDoc (`# Excluding auto-copy defaults` section, `@param excludes`)
- `cream-ksp` core:
  - `KSAnnotationExt.extractExcludes()` reads the raw `excludes` argument
  - `GenerateSourceAnnotation.CopyMapping.excludes` (with reverse-direction translation) / `GenerateSourceAnnotation.CombineMapping.excludes`
  - `ExcludeProperty.isExcludedFromCopy` handles the mapping branches by generated-parameter name; new `warnIfMappingExcludesHaveNoEffect` for the unmatched warning
  - `copyFun/Class.kt` and `combineFun/CombineToClass.kt` invoke the warning once per generated function set
- Snapshot tests: new `excludes` families appended last (`13--excludes` for CopyMapping — 5 scenarios incl. canReverse and Map+canReverse; `12--excludes` for CombineMapping — 3 scenarios incl. the unmatched-name warning). 32 new goldens; **no existing golden churn**.
- E2E (`test/` module): `copyMapping/Exclude(.Test).kt` (3 cases incl. reverse-direction translated exclude) and `combineMapping/Exclude(.Test).kt` (1 case), following the existing mapping E2E precedent.
- Docs: `doc/customization/exclude.md` / `.ja.md` (placement table rows, new `CopyMapping / CombineMapping — excludes` section with the canReverse rule, edge-case note), `doc/internal/feature-and-options.md` `.Exclude` row cells (`✕` → `✓ excludes`, only those two cells), README/README.ja customization table cell.

## Test plan

- [x] `./gradlew ktlintCheck` — BUILD SUCCESSFUL
- [x] `./gradlew :cream-ksp:test` — BUILD SUCCESSFUL (1592 tests, 0 failures; clean verifying pass after `-Dcream.snapshot.update=true` regeneration; no existing goldens modified)
- [x] `./gradlew :test:jvmTest` — BUILD SUCCESSFUL (150 tests, 0 failures, incl. the 4 new E2E cases)

Closes #164

> Stacked on #176 ← #175 (merge order: #175 → #176 → this PR).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: cream-ksp/src/main/kotlin/me/tbsten/cream/ksp/core/copyFun/Class.kt [hunk 1]
```diff
@@ -32,6 +33,15 @@ internal fun Appendable.appendCopyToClassFunction(
...
     omitPackages: List<String>,
 ) {
+    // Warn once per (source, target) pair: an entry is effective when it suppresses a `= this.x`
+    // default in ANY generated overload (any constructor).
+    warnIfMappingExcludesHaveNoEffect(
+        targetParameters = targetClass.getConstructors().flatMap { it.parameters }.toList(),
+        sources = listOf(source),
+        generateSourceAnnotation = generateSourceAnnotation,
+        logger = logger,
+    )
```
[the remark this answers is not in the document]
Comment (TBSten) [reply cream-ksp/src/main/kotlin/me/tbsten/cream/ksp/core/copyFun/Class.kt]: Fixed in 3f0bf184 — hoisted the warning out of the per-function core generators into the copyMapping / combineMapping features, evaluated once per mapping annotation across both directions: no duplicate warning under canReverse, the message always names the user-written entry (never the Map-translated reverse spelling), an entry effective in only one direction is no longer a false positive, sealed fan-out warns once over the union of leaves, and object targets (which generate no parameters) now warn too. Locked by the new snapshot scenario excludeUnmatchedNameWithCanReverse (exactly one warning).

File: cream-ksp/src/main/kotlin/me/tbsten/cream/ksp/core/common/ExcludeProperty.kt [hunk 2]
```diff
@@ -124,3 +129,68 @@ internal fun KSPropertyDeclaration.warnIfSourceExcludeHasNoEffect(
...
+ * mapping holder because the mapped external classes carry no annotation to point at.
+ *
+ * Evaluated once per mapping annotation (called from the mapping features, not per generated
+ * function), so neither the `canReverse` reverse pass nor a sealed target's per-leaf fan-out can
+ * duplicate the warning, and the message always names the entry as the user wrote it
+ * ([originalExcludes]) — never its Map-translated reverse spelling. An entry is effective when
+ * its direction-local spelling suppresses a default on any constructor parameter of any concrete
+ * target in any direction (every transitive concrete leaf when the target is sealed; an `object`
+ * target has no parameters, so its entries always warn).
+ */
```
[the remark this answers is not in the document]
Comment (TBSten) [reply cream-ksp/src/main/kotlin/me/tbsten/cream/ksp/core/common/ExcludeProperty.kt:159]: Aligned the PR description with the implementation (the description was stale). The implemented behavior is intentional: since 729f540c the warning is evaluated once per mapping annotation, and an object target generates a parameterless function, so its excludes entries can never suppress a default — they always warn, per the issue #164 spec ("an entry that does not match any auto-defaulted parameter has no effect and emits a KSP warning"). The KDoc here already documents this; the PR body now matches.
